MatrixElement symbol property added; validate shape compatible with indices

smichr opened the pull request:
#### Brief description of what is fixed or changed

fixes #22676

##### symbol attribute has been added
MatrixElement behaves more like Indexed in terms of argument access:

```python
>>> IndexedBase('x')[0].base
x
>>> MatrixSymbol('x',2,2)[0,0].symbol
x
```

##### invalid substitutions no longer succeed

In master,
```python
X = MatrixSymbol("X", 2, 2)
Y = MatrixSymbol("Y", 1, 2)
assert X[1, 1].subs(X, Y) == Y[1, 1]
>>> Y[1,1]
Traceback (most recent call last):
...
IndexError: indices out of bounds
```
in this PR that substitution no longer succeeds so you don't end up with an invalid expression

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->


#### Other comments

There is some divergence in how MatrixSymbol/IndexedBase and MatrixElement/Indexed behave in terms of free symbols and attributes. This should have a look by someone that works with matrix expressions to see if these changes make sense.

- [x] QUESTION: is there are reason to store Indexed.label as a Symbol but MatrixSymbol as a Str?
ANSWER: no (see #22391 and #22272)

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* matrices
    * replacement of MatrixSymbol in MatrixElement requires shape to be compatible with indices to succeed
    * MatrixElement now has `symbol` attributes (analogous to `base` attribute of `IndexedBase` )
    * `eval_hyper_sum` returns as unchanged arguments for which `is_hypergeometric` is False and those which produce singularities at `a` or `b + 1` where `and` and `b` are the lower and upper limits of evaluation, respectively.
* stats
    * removed deprecation warning about using `evaluate=False`
<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -613,6 +621,17 @@ def doit(self, **kwargs):
     def indices(self):
         return self.args[1:]
 
+    def _eval_subs(self, old, new):
+        # when replacing the symbol, the indices cannot
+        # access out of bounds
+        if old == self.symbol:
+            if type(old) == type(new):
+                for i, j in zip(self.indices, new.shape):
+                    if i.is_Integer and j.is_Integer and i >= j:
+                        raise IndexError('indices access out of range')
```
I think it would be better to check this in `MatrixElement.__new__` rather than `_eval_subs`.

smichr replied:
I tried and failed but I think I got thrown off by thinking the MatrixBase clause was for MatrixSymbol...I will try again to make the change there.

smichr replied:
OK, this is part of `__new__` now.

smichr commented:
@oscarbenjamin , any suggestions for working around mypy errors? Do I just put "  # type: ignore" on the class def?

oscarbenjamin commented:
> @oscarbenjamin , any suggestions for working around mypy errors? Do I just put " # type: ignore" on the class def?

I think that these errors are really pointing to bugs in the code somewhere but it's not obvious where those are. It's also possible that they are due to incorrect type-hints being applied somewhere. It should always be possible to silence mypy by adding `type: ignore` on the exact line number that is referenced in the mypy output.

smichr committed:
MatrixElement gets name and symbol properties

smichr committed:
check indices

smichr committed:
no Trace in hyper_sum; conv mut to immut

smichr committed:
remove name attribute

smichr commented:
@Upabjojr @oscarbenjamin , does this look ok?

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1308,7 +1308,10 @@ def eval_sum_hyper(f, i_a_b):
+            try:
+                res2 = _eval_sum_hyper(f, i, b + 1)
+            except IndexError:  # for f = Trace (or other matexpr?)
+                res2 = None
```
Where is the IndexError coming from? It doesn't seem like a reasonable exception to catch here.

smichr replied:
`Trace(A).rewrite(Sum)` -- the `hyper_sum` method is attempted and that method requires sampling at `upper limit + 1` which you cannot do with a physical object like a matrix.

oscarbenjamin replied:
I don't quite follow how `hyper_sum` ends up being called. Presumably this is a case where it shouldn't be called.

smichr replied:
caught at source, now

oscarbenjamin commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -589,10 +590,15 @@ class MatrixElement(Expr):
+        elif isinstance(name, MatrixSymbol):
+            r, c = name.shape
+            if is_le(r, n) or is_le(c, m):
+                raise IndexError('indices out of bounds')
```
What about negative indices?

oscarbenjamin commented on sympy/tensor/array/expressions/arrayexpr_derivatives.py:
```diff
@@ -23,7 +23,7 @@ def array_derive(expr, x):
 @array_derive.register(Expr) # type: ignore
 def _(expr: Expr, x: Expr):
-    return ZeroArray(*x.shape)
+    return ZeroArray(*x.shape)  # type: ignore
```
I would rather fix these things than add `ignore`. This is clearly invalid because an arbitrary Expr does not have a `.shape` attribute.

smichr replied:
This must be a change to mypy. It couldn't be a PR that was added or else that PR wouldn't have passed tests. And this happening on all PRs. So we either patch it until it gets fixed or wait until someone can fix it and stop all other PRs. What is your preference?

oscarbenjamin replied:
The problem was initiated by a change in mypy. I thought that was now addressed (in the ignore sense) but the diff here shows this as a new change in SymPy.

smichr replied:
#22692 is passing and has the following difference to master (and did not add those differences, to it was a change to master that did this):
```diff
diff --git a/.github/workflows/runtests.yml b/.github/workflows/runtests.yml
index 132075145e..f9c6215388 100644
--- a/.github/workflows/runtests.yml
+++ b/.github/workflows/runtests.yml
@@ -25,7 +25,7 @@ jobs:
         with:
           python-version: '3.9'
       - run: python -m pip install --upgrade pip
-      - run: pip install mpmath flake8
+      - run: pip install mpmath flake8 mypy
 
       - name: Basic code quality tests
         run: bin/test quality
@@ -33,6 +33,9 @@ jobs:
       - name: Run flake8 on the sympy package
         run: flake8 sympy
 
+      - name: Run mypy on the sympy package
+        run: mypy sympy
+
       - name: Detect invalid escapes like '\e'
         run: python -We:invalid -We::SyntaxWarning -m compileall -f -q sympy/
 
@@ -49,28 +52,6 @@ jobs:
       #- name: Make sure all commits have an associated author
       #  run: bin/authors_update.py
 
-
-  # ----------------------------- mypy ----------------------------- #
-
-  mypy:
-    needs: code-quality
-
-    runs-on: ubuntu-20.04
-    steps:
-      - uses: actions/checkout@v2
-        with:
-          # Clone full git history (needed for AUTHORS/mailmap)
-          fetch-depth: '0'
-
-      - uses: actions/setup-python@v2
-        with:
-          python-version: '3.9'
-      - run: python -m pip install --upgrade pip
-      - run: pip install mpmath mypy
-
-      - name: Run mypy on the sympy package
-        run: mypy sympy
-
```

smichr replied:
@asmeurer , see previous. I think the additions of the red and removal of green diff lines is what made master start failing mypy.

smichr committed:
use valid_index to test subs; allow neg indices

smichr committed:
don't use doit on MatExpr

oscarbenjamin commented on sympy/matrices/expressions/trace.py:
```diff
@@ -144,8 +145,11 @@ def get_arg_key(x):
+        i = uniquely_named_symbol('i', expr)
+        s = Sum(self.arg[i, i], (i, 0, self.arg.rows - 1))
+        if not isinstance(expr, MatrixExpr):
+            s = s.doit()
```
If this call to `doit` would otherwise raise `IndexError` then that suggests that something should still be fixed somewhere because `doit` shouldn't raise `IndexError`.

smichr replied:
Have a look at the `eval_sum_hyper` routine. It needs to access the `upper bound + 1`: `res2 = _eval_sum_hyper(f, i, b + 1)`. This will fail for any indexed object which cannot be computed at that bound. I left a note on the routine regarding indexed objects and I added a safeguard in the trace routine that is writing trace of a matrix as a sum. In the case of an actual matrix the dimensions are concrete and `eval_sum_hyper` is not called.

oscarbenjamin replied:
Can you show example code needed to set this up so I can see the error? Would I see the error on master or only with the changes in this PR?

smichr replied:
Saying that it shouldn't happen is equivalent to saying either "do routine writing sums with indexed objects should call the `eval_sum_hyper` routine. But calling that routine is part of the `Sum.doit` machinery. So any routine writing Sum with indexed objects should not call doit. And that is the change that I made in `trace`.

oscarbenjamin replied:
What I was looking for was how to reproduce this. I see now that with this PR it's:
```python
In [1]: n = Symbol('n', integer=True)

In [2]: M = MatrixSymbol('M', n, n)

In [3]: trace(M).rewrite(Sum).doit()
---------------------------------------------------------------------------
IndexError 
```
Strangely the `integer=True` is needed here.

As you say it goes wrong from here:
https://github.com/sympy/sympy/blob/14cf005b8fea132207654603cc6495900040106a/sympy/concrete/summations.py#L1310-L1311
There could be lots of situations where just substituting `b+1` in might break something so either the exception needs to be caught here (i.e. your previous code) or there should be some code to determine whether this is an appropriate function to use in this routine at all. There isn't much documentation but I presume that `eval_sum_hyper` is only supposed to work for hypergeometric functions and `M[i, i]` is clearly not hypergeometric in `i`.

oscarbenjamin replied:
Catching the exception is not exactly satisfactory because this kind of potentially invalid operation could lead to all sorts of different exceptions.

oscarbenjamin replied:
These lines seem to be responsible for this output for example:
```python
In [1]: b = symbols('b', integer=True, positive=True)

In [2]: summation(1/(x-b)**2, (x, 0, b-1))
Out[2]: 
 ⎛         ⅈ⋅π⎞      
Φ⎝1, 2, b⋅ℯ   ⎠ + zoo
```

oscarbenjamin replied:
This is giving zoo in cases where actually the result might be well defined:
```python
In [7]: Sum(1/(x-b)**2, (x, 0, b-1)).subs(b, 2).doit()
Out[7]: 5/4

In [8]: b = symbols('b', integer=True, positive=True)

In [9]: s = Sum(1/(x-b)**2, (x, 0, b-1))

In [10]: s.subs(b, 2).doit()
Out[10]: 5/4

In [11]: s.doit().subs(b, 2)
Out[11]:
---------------------------------------------------------------------------
ZeroDivisionError  
```

smichr replied:
I added a `is_hypergeometric` tests to the routine and then a check that illegals are not introduced by the `b + 1` evaluation. From the wolframalpha result, it appears that a solution is possible. That's a separate issue; at least we can check that there is no `zoo`.

oscarbenjamin replied:
I'm not sure how well `is_hypergeometric` works. Note that it just calls `hypersimp` which is already being called here I think.

smichr replied:
bad evaluations are avoid now and tested as:
```python
def test_pr_22677():
    b = Symbol('b', integer=True, positive=True)
    assert Sum(1/x**2,(x, 0, b)).doit() == Sum(x**(-2), (x, 0, b))
    assert Sum(1/(x - b)**2,(x, 0, b-1)).doit() == Sum(
        (-b + x)**(-2), (x, 0, b - 1))
```
The first one gives the following in master:
```python
>>> Sum(1/x**2,(x,0,b)).doit()
zoo - (b**2 + 2*b + 1)*lerchphi(1, 2, b + 1)/(b + 1)**2
```

smichr committed:
add concrete test

smichr committed:
do hypergeometric check

smichr committed:
relax is_hypergeometric check; remove warn; update

smichr committed:
watch for illegals at both limits

oscarbenjamin commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -589,10 +589,14 @@ class MatrixElement(Expr):
-        if isinstance(name, str):
+            name = _sympify(name)  # change mutable into immutable
+        elif isinstance(name, MatrixSymbol):
+            if not name.valid_index(n, m):
```
I think that `valid_index` should always be checked, not just for MatrixSymbol e.g.:
```python
In [18]: n = Symbol('n', integer=True, positive=True)

In [19]: M = MatrixSymbol('M', n, n)

In [20]: M[3, m]
Out[20]: (M)[3, m]

In [21]: M[3, m].subs(M, Matrix([[1, 2], [3, 4]]))
Out[21]: 
⎛⎡1  2⎤⎞      
⎜⎢    ⎥⎟[3, m]
⎝⎣3  4⎦⎠ 
```

smichr replied:
good point! To me this seems like a strange constructor in that you are creating an unevaluated indexed object rather than simply indexing an object. It works, but having to put these checks here seems odd to me.

smichr committed:
use valid_index on all name

smichr committed:
only check if obj has valid_index

>>> from sympy.stats import MultivariateNormal, density
>>> Sg = MatrixSymbol('Sg', n, n)
>>> mu = MatrixSymbol('mu', n, 1)
>>> X = MultivariateNormal('X', mu, Sg)
>>> obs = MatrixSymbol('obs', n, 1)
>>> density(X)(obs)
exp(((1/2)*mu.T - (1/2)*obs.T)*Sg**(-1)*(-mu + obs))/
sqrt((2*pi)**n*Determinant(Sg))

This is a Mul: exp(1x1)/scalar. It is a MatrixKind(NumberKind)
but is not a MatMul or MatExpr and does not have a valid_index
method.

smichr committed:
restore 15943
